### PR TITLE
Create controlled greater than bloq

### PR DESCRIPTION
**Pull request by NoureldinYosri**
A controlled version of greater than is needed in order to be able to create controlled mod addition gate.

In this PR I create CLinearDepthGreaterThan and fix some classical and QDType bugs in other bloqs

I also update OnEach to accept a target dtype ... that's because OnEach returns a register of type QAny which causes problems in classical simulation. this way we can expect maintain the dtype of the input register.

---

I updated https://github.com/quantumlib/Qualtran/issues/1278 so that it tracks all .controlled work

### CODE REVIEW

**Comment by fdmalone** [comment]
Incidentally fixes #1230

File: qualtran/bloqs/arithmetic/comparison.py [hunk 1]
```diff
@@ -1017,3 +1024,135 @@ def _eq_k() -> EqualsAConstant:
 
 
 _EQUALS_K_DOC = BloqDocSpec(bloq_cls=EqualsAConstant, examples=[_eq_k])
+
+
+@frozen
+class CLinearDepthGreaterThan(Bloq):
+    r"""Controlled greater than between two integers.
+
+    Implements |c>|a>|b>|t> => |a>|b>|t ⨁ ((a > b)c)> using $n+2$ Toffoli gates.
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/comparison.py:1033]: nit: use Latex in docstrings?

File: qualtran/bloqs/arithmetic/comparison.py [hunk 2]
```diff
@@ -1017,3 +1024,135 @@ def _eq_k() -> EqualsAConstant:
...
+
+    Note: the true cost is $n+1$ but an extra Toffoli comes from OutOfPlaceAdder which operates
+    on $n+1$ qubits rather than $n$. Changing the definition of OutOfPlaceAdder will remove this
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/comparison.py:1036]: What does changing the definition mean here? Is this a bug, maybe open an issue and link?

File: qualtran/bloqs/arithmetic/comparison.py (hunk 1 above)
Comment (fdmalone) [reply qualtran/bloqs/arithmetic/comparison.py:1033]: The RHS of the => seems to be missing the $|c\rangle$ register before $|a\rangle$

File: qualtran/bloqs/arithmetic/comparison.py (hunk 2 above)
Comment (NoureldinYosri) [reply qualtran/bloqs/arithmetic/comparison.py:1036]: not a bug just a different meaning ... the outofplaceadder adds to $n$ bit ints into an $n+1$ register ... so the addition is exact ... other bloqs like Add and Subtract do the operation into just $n$ bits (so mod $2^n$)

File: qualtran/bloqs/arithmetic/comparison.py [hunk 3]
```diff
@@ -1017,3 +1024,135 @@ def _eq_k() -> EqualsAConstant:
...
+        dtype: type of the integer registers.
+        cv: ctrl value at which the bloq is active.
+
+    Registers:
+        a: dtype input registers.
+        b: dtype input registers.
+        target: A single bit output register to store the result of a > b.
+
+    References:
+        [Halving the cost of quantum addition](https://arxiv.org/abs/1709.06648).
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/comparison.py:1053]: nit: typically add more context (section 2, page 4, Eq 7 etc)

File: qualtran/bloqs/arithmetic/comparison.py [hunk 4]
```diff
@@ -1017,3 +1024,135 @@ def _eq_k() -> EqualsAConstant:
...
+class CLinearDepthGreaterThan(Bloq):
+    r"""Controlled greater than between two integers.
+
+    Implements |c>|a>|b>|t> => |a>|b>|t ⨁ ((a > b)c)> using $n+2$ Toffoli gates.
+
+    Note: the true cost is $n+1$ but an extra Toffoli comes from OutOfPlaceAdder which operates
+    on $n+1$ qubits rather than $n$. Changing the definition of OutOfPlaceAdder will remove this
+    extra Toffoli.
+
+    This comparator relies on the fact that (b' + a)' = b - a. If a > b, then b - a < 0. We
```
[the remark this answers is not in the document]
Comment (NoureldinYosri) [reply qualtran/bloqs/arithmetic/comparison.py:1039]: bitwise not (I.e. `~x`)

File: qualtran/bloqs/arithmetic/comparison.py [hunk 5]
```diff
@@ -1017,3 +1024,135 @@ def _eq_k() -> EqualsAConstant:
...
+    def signature(self) -> Signature:
+        return Signature.build_from_dtypes(ctrl=QBit(), a=self.dtype, b=self.dtype, target=QBit())
+
+    def wire_symbol(
+        self, reg: Optional['Register'], idx: Tuple[int, ...] = tuple()
+    ) -> 'WireSymbol':
+        if reg is None:
+            return Text('')
+        if reg.name == 'ctrl':
+            return TextBox('c')
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/comparison.py:1071]: Are we not doing Circles for controls in the diagrams?

File: qualtran/bloqs/arithmetic/comparison.py (hunk 3 above)
Comment (NoureldinYosri) [reply qualtran/bloqs/arithmetic/comparison.py:1053]: will do ... I just copied the docstring of LinearDepthGreaterThan

File: qualtran/bloqs/arithmetic/comparison.py [hunk 6]
```diff
@@ -1017,3 +1024,135 @@ def _eq_k() -> EqualsAConstant:
...
+        if reg.name == "a":
+            return TextBox('a')
+        if reg.name == "b":
+            return TextBox('b')
+        if reg.name == "target":
+            return TextBox('t⨁((a>b)c)')
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/comparison.py:1077]: Double check the qpic conversion https://github.com/quantumlib/Qualtran/blob/fe32dfd8a71bd8c769da1dd8572a6666141f2c1e/qualtran/drawing/qpic_diagram.py#L49 to make sure the > is not unicode (here or there, update either appropriately). --  Never mind I think this is fine.

Comment (fdmalone) [reply qualtran/bloqs/arithmetic/comparison.py:1077]: I was thinking of ^ which rendered weird before.

File: qualtran/bloqs/arithmetic/comparison.py [hunk 7]
```diff
@@ -1017,3 +1024,135 @@ def _eq_k() -> EqualsAConstant:
...
+        dtype = attrs.evolve(self.dtype, bitsize=self.dtype.bitsize + 1)
+        b = bb.add(BitwiseNot(dtype), x=b)  # b := -b-1
+        a = bb.add(Cast(dtype, QUInt(dtype.bitsize)), reg=a)
+        b = bb.add(Cast(dtype, QUInt(dtype.bitsize)), reg=b)
+        a, b, c = bb.add(OutOfPlaceAdder(self.dtype.bitsize + 1), a=a, b=b)  # c := a - b - 1
+        c = bb.add(BitwiseNot(QUInt(dtype.bitsize + 1)), x=c)  # c := b - a
+
+        # Update `target`
+        c_arr = bb.split(c)
+        (ctrl_q,) = bb.split(ctrl)
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/comparison.py:1100]: Is this split necessary?

File: qualtran/bloqs/arithmetic/comparison.py [hunk 8]
```diff
@@ -1017,3 +1024,135 @@ def _eq_k() -> EqualsAConstant:
...
+        (ctrl_q, c_arr[1]), target = bb.add(
```
Comment (fdmalone) [inline qualtran/bloqs/arithmetic/comparison.py:1101]: Question: why is it bit 1?

**Review by fdmalone — CHANGES_REQUESTED** [review]
Some nits / questions.

File: qualtran/bloqs/arithmetic/comparison.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fdmalone) [reply qualtran/bloqs/arithmetic/comparison.py:1039]: I think just use ~ which is easier to read maybe?

File: qualtran/bloqs/arithmetic/comparison.py (hunk 5 above)
Comment (fdmalone) [reply qualtran/bloqs/arithmetic/comparison.py:1071]: For example: https://github.com/quantumlib/Qualtran/blob/fe32dfd8a71bd8c769da1dd8572a6666141f2c1e/qualtran/bloqs/factoring/ecc/ec_add_r.py#L82

**Comment by fdmalone** [comment]
Nice!

File: qualtran/bloqs/arithmetic/comparison.py (hunk 1 above)
Comment (NoureldinYosri) [reply qualtran/bloqs/arithmetic/comparison.py]: thanks for catching this

File: qualtran/bloqs/arithmetic/comparison.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (NoureldinYosri) [reply qualtran/bloqs/arithmetic/comparison.py]: updated

File: qualtran/bloqs/arithmetic/comparison.py (hunk 5 above)
Comment (NoureldinYosri) [reply qualtran/bloqs/arithmetic/comparison.py]: done

File: qualtran/bloqs/arithmetic/comparison.py (hunk 7 above)
Comment (NoureldinYosri) [reply qualtran/bloqs/arithmetic/comparison.py]: removed

File: qualtran/bloqs/arithmetic/comparison.py (hunk 8 above)
Comment (NoureldinYosri) [reply qualtran/bloqs/arithmetic/comparison.py]: the 0th bit is the exta bit that OutOfPlaceAdder added which in this case indicates an overflow or not... I added a comment explaining this

**Review by fdmalone — APPROVED** [review]

**Pull request merged**